File: DSAjavascript/DynamicPrograming.js/CombinationofSum.js
/**
 * 
 * leetcode problem no: 337
 * 
 * leetcode solution link(https://leetcode.com/problems/combination-sum-iv/description/)
 */

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */

let arr;
let dp;

function f ( target){

    //Base case

    if ( target == 0)
    return 1;

    if (dp[target ] != -1){
        return dp[target];
    }

  let  result = 0;
  let n = arr.length;
    for(let  k  = 0; k<n ; k++)
       if( target - arr[k] >= 0){

        result  += f(target - arr[k]);
         
       }

      return dp[target] = result;


}
var combinationSum4 = function(nums, target) {
  arr = nums;
    dp = Array(1005).fill(-1);
    
     return f(target);
};

File: DSAinJava/DynamicPrograming.java/CombinationofSum.java
/**
 * leetcode problem no : 337
 * 
 * leetcode solution link (https://leetcode.com/problems/combination-sum-iv/description/)
 */

import java.util.Arrays;

public class CombinationofSum {

       public static int[] arr;
     public static int[] dp = new int[1005];

    public static int f (int target){

        //Base case

        if ( target == 0)
        return 1;
    
        if (dp[target ] != -1){
            return dp[target];
        }

      int  result = 0;
      int n = arr.length;
        for(int k  = 0; k<n ; k++)
           if( target - arr[k] >= 0){
  
            result  += f(target - arr[k]);
             
           }

          return dp[target] = result;


    }
    public int combinationSum4(int[] nums, int target) {
         arr = nums;
           Arrays.fill(dp , -1);  
        
         return f(target);
    }
    
}
